File: redis-proxy-dataplane-java/src/main/java/com/example/redisproxy/dataplane/governance/NamespaceLimiter.java
package com.example.redisproxy.dataplane.governance;

import com.example.redisproxy.dataplane.config.ProxyProperties;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Tag;
import java.time.Clock;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.springframework.stereotype.Component;

@Component
public class NamespaceLimiter {
    private final MeterRegistry registry;
    private final Map<String, Integer> connections = new HashMap<>();
    private final Map<String, Integer> inflight = new HashMap<>();
    private final Map<String, QpsWindow> qpsWindows = new HashMap<>();
    private final Map<String, AtomicInteger> connectionGauges = new HashMap<>();
    private final Map<String, AtomicInteger> inflightGauges = new HashMap<>();
    private Clock clock = Clock.systemUTC();

    public NamespaceLimiter(MeterRegistry registry) {
        this.registry = registry;
    }

    public synchronized LimitResult bind(String current, ProxyProperties.Namespace next) {
        if (next == null || next.getName() == null || next.getName().isBlank() || next.getName().equals(current)) {
            return LimitResult.allow();
        }
        String namespace = next.getName();
        if (next.getLimits().getMaxConnections() > 0 && connections.getOrDefault(namespace, 0) >= next.getLimits().getMaxConnections()) {
            return LimitResult.rejected("connection_limit");
        }
        if (current != null && !current.isBlank() && connections.getOrDefault(current, 0) > 0) {
            connections.put(current, connections.get(current) - 1);
            gauge(connectionGauges, "redis.proxy.namespace.connections", current).set(connections.get(current));
        }
        connections.put(namespace, connections.getOrDefault(namespace, 0) + 1);
        gauge(connectionGauges, "redis.proxy.namespace.connections", namespace).set(connections.get(namespace));
        return LimitResult.allow();
    }

    public synchronized void unbind(String namespace) {
        if (namespace == null || namespace.isBlank()) {
            return;
        }
        if (connections.getOrDefault(namespace, 0) > 0) {
            connections.put(namespace, connections.get(namespace) - 1);
        }
        gauge(connectionGauges, "redis.proxy.namespace.connections", namespace).set(connections.getOrDefault(namespace, 0));
    }

    public synchronized LimitResult allowRequest(ProxyProperties.Namespace namespace) {
        if (namespace == null || namespace.getName() == null || namespace.getName().isBlank()) {
            return LimitResult.allow();
        }
        String name = namespace.getName();
        ProxyProperties.NamespaceLimits limits = namespace.getLimits();
        long second = clock.millis() / 1000;
        if (limits.getMaxQps() > 0) {
            QpsWindow window = qpsWindows.get(name);
            if (window == null || window.second != second) {
                window = new QpsWindow(second, 0);
            }
            if (window.count >= limits.getMaxQps()) {
                qpsWindows.put(name, window);
                return LimitResult.rejected("qps_limit");
            }
            qpsWindows.put(name, new QpsWindow(second, window.count + 1));
        }
        if (limits.getMaxInflight() > 0 && inflight.getOrDefault(name, 0) >= limits.getMaxInflight()) {
            return LimitResult.rejected("inflight_limit");
        }
        inflight.put(name, inflight.getOrDefault(name, 0) + 1);
        gauge(inflightGauges, "redis.proxy.namespace.inflight", name).set(inflight.get(name));
        return LimitResult.allow();
    }

    public synchronized void finishRequest(String namespace) {
        if (namespace == null || namespace.isBlank()) {
            return;
        }
        if (inflight.getOrDefault(namespace, 0) > 0) {
            inflight.put(namespace, inflight.get(namespace) - 1);
        }
        gauge(inflightGauges, "redis.proxy.namespace.inflight", namespace).set(inflight.getOrDefault(namespace, 0));
    }

    void setClock(Clock clock) {
        this.clock = clock;
    }

    private AtomicInteger gauge(Map<String, AtomicInteger> gauges, String name, String namespace) {
        return gauges.computeIfAbsent(namespace, key ->
                registry.gauge(name, List.of(Tag.of("namespace", key)), new AtomicInteger()));
    }

    public record LimitResult(boolean allowed, String reason) {
        static LimitResult allow() {
            return new LimitResult(true, "");
        }

        static LimitResult rejected(String reason) {
            return new LimitResult(false, reason);
        }
    }

    private record QpsWindow(long second, int count) {
    }
}

File: redis-proxy-dataplane-go/internal/proxy/namespace_limiter.go
package proxy

import (
	"sync"
	"time"

	"github.com/example/redis-proxy-dataplane-go/internal/config"
	"github.com/example/redis-proxy-dataplane-go/internal/metrics"
)

type namespaceLimiter struct {
	mu         sync.Mutex
	reg        *metrics.Registry
	conns      map[string]int
	inflight   map[string]int
	qpsWindows map[string]qpsWindow
}

type qpsWindow struct {
	second int64
	count  int
}

func newNamespaceLimiter(reg *metrics.Registry) *namespaceLimiter {
	return &namespaceLimiter{
		reg:        reg,
		conns:      map[string]int{},
		inflight:   map[string]int{},
		qpsWindows: map[string]qpsWindow{},
	}
}

func (l *namespaceLimiter) bind(current string, next config.NamespaceConfig) (bool, string) {
	if next.Name == "" || current == next.Name {
		return true, ""
	}
	l.mu.Lock()
	defer l.mu.Unlock()
	if next.Limits.MaxConnections > 0 && l.conns[next.Name] >= next.Limits.MaxConnections {
		return false, "connection_limit"
	}
	if current != "" && l.conns[current] > 0 {
		l.conns[current]--
		l.reg.NamespaceConnections.WithLabelValues(current).Set(float64(l.conns[current]))
	}
	l.conns[next.Name]++
	l.reg.NamespaceConnections.WithLabelValues(next.Name).Set(float64(l.conns[next.Name]))
	return true, ""
}

func (l *namespaceLimiter) unbind(namespace string) {
	if namespace == "" {
		return
	}
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.conns[namespace] > 0 {
		l.conns[namespace]--
	}
	l.reg.NamespaceConnections.WithLabelValues(namespace).Set(float64(l.conns[namespace]))
}

func (l *namespaceLimiter) allowRequest(namespace config.NamespaceConfig) (bool, string) {
	if namespace.Name == "" {
		return true, ""
	}
	now := time.Now().Unix()
	l.mu.Lock()
	defer l.mu.Unlock()
	if namespace.Limits.MaxQPS > 0 {
		window := l.qpsWindows[namespace.Name]
		if window.second != now {
			window = qpsWindow{second: now}
		}
		if window.count >= namespace.Limits.MaxQPS {
			l.qpsWindows[namespace.Name] = window
			return false, "qps_limit"
		}
		window.count++
		l.qpsWindows[namespace.Name] = window
	}
	if namespace.Limits.MaxInflight > 0 && l.inflight[namespace.Name] >= namespace.Limits.MaxInflight {
		return false, "inflight_limit"
	}
	l.inflight[namespace.Name]++
	l.reg.NamespaceInflight.WithLabelValues(namespace.Name).Set(float64(l.inflight[namespace.Name]))
	return true, ""
}

func (l *namespaceLimiter) finishRequest(namespace string) {
	if namespace == "" {
		return
	}
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.inflight[namespace] > 0 {
		l.inflight[namespace]--
	}
	l.reg.NamespaceInflight.WithLabelValues(namespace).Set(float64(l.inflight[namespace]))
}
